sdc/rewrites/dataframe_constructor.py

```diff
@@ -302,20 +308,76 @@ def pd_dataframe_overload(data, index=None, columns=None, dtype=None, copy=False
+            or isinstance(data, types.Array) and data.ndim == 2 and isinstance(data.dtype, types.Number)):
+        ty_checker.raise_exc(data, 'dict or 2d numeric array', 'data')
 
-    if not (isinstance(index, (types.Omitted, types.ListType, types.List,
-                               types.Array, StringArrayType, types.NoneType) or index is None)):
+    accepted_index_types = (types.Omitted, types.NoneType, types.ListType, types.List) + sdc_pandas_index_types
+    if not (isinstance(index, accepted_index_types) or index is None):
         ty_checker.raise_exc(index, 'array-like', 'index')
 
-    if not (isinstance(columns, (types.Omitted, types.NoneType, types.Tuple, types.UniTuple) or columns is None)):
+    if not (isinstance(columns, (types.Omitted, types.NoneType, types.Tuple, types.UniTuple)) or columns is None):
         ty_checker.raise_exc(columns, 'tuple of strings', 'columns')
 
-    if not (isinstance(dtype, (types.Omitted, types.NoneType) or dtype is None)):
+    if not (isinstance(dtype, (types.Omitted, types.NoneType)) or dtype is None):
         ty_checker.raise_exc(dtype, 'None', 'dtype')
 
-    if not (isinstance(copy, (types.Omitted, types.NoneType) or columns is False)):
+    if not (isinstance(copy, (types.Omitted, types.NoneType)) or copy is False):
         ty_checker.raise_exc(copy, 'False', 'copy')
 
+    if isinstance(data, types.Array):
+        # case of homogenous DF columns, is special as we can use views to input data
+        # when creating internal DF structure and avoid penalty in boxing DF as
+        # pd.DataFrame can be created from 2d array without copy
+
+        nb_col_names = None
+        try:
+            nb_col_names = tuple(get_nbtype_literal_values(columns))
+        except AssertionError:
+            ty_checker.raise_exc(columns, 'tuple of literal strings', 'columns')
+
+        n_cols = len(columns)
+        if index is None and n_cols == 0:
+            # DataFrame index type cannot be defined unless columns argument is provided
+            # as it depends on the runtime number of columns in data
+            raise SDCLimitation("pd.DataFrame constructor from np.ndarray " \
+                                f"requires columns argument. Given columns={columns}.")
+
+        # FIXME: there should be more accurate way to write this layout definition
+        if data.layout in ('C', 'A'):
+            col_type = types.Array(data.dtype, 1, 'A')
+        else:
+            col_type = types.Array(data.dtype, 1, 'C')
+        nb_col_types = tuple([col_type, ] * n_cols)
+        column_loc, _, _ = get_structure_maps(nb_col_types, nb_col_names)
+
+        typingctx = resolve_dispatcher_from_str(current_target()).targetdescr.typing_context
+        fnty = typingctx.resolve_value_type(fix_df_index)
+        nb_index_type = types.none if index is None else index
+        fixed_index_sig = fnty.get_call_type(typingctx, (nb_index_type, nb_col_types[0]), {})  ### FIXME: need add column argument
```

Comment:
@kozlov-alexey , hi. I reference this part of code to generate the output `dataframetype` for my jit version of `dataframe.apply()`. But, as you noted, how to specify the column names that users provide?

My code is as below (My assumption is that users' `udf` will return a `pd.Series` and provide the `real_col_names` (a python `str tuple`) as the output dataframe column. Now, I generate a random `fake_col_names` like `(c1, c2, c3, ...)` with the same length as `real_col_names`):
```
# Referenced from:
# https://gist.github.com/kozlov-alexey/f29e8d2703789491e8e24e41de16536b
# https://github.com/IntelPython/sdc/pull/1000/files
# limitations:
# 1. could not point out col_names
@sdc_overload_method(DataFrameType, 'apply')
def pd_dataframe_apply_overload(df, udf, args, real_col_names):
    """generate output dataframe type for impl"""
    col_len_outside = len(real_col_names)
    fake_col_names = tuple(['c' + str(i + 1) for i in range(col_len_outside)])
    col_type = types.Array(types.float64, 1, 'C')
    col_types = tuple([col_type, ] * col_len_outside)
    column_loc, _, _ = get_structure_maps(col_types, fake_col_names)
    typingctx = resolve_dispatcher_from_str(
        current_target()
    ).targetdescr.typing_context
    fnty = typingctx.resolve_value_type(fix_df_index)
    index_type = df.index

    """ fixme(from kov): add column argument"""
    fixed_index_sig = fnty.get_call_type(
        typingctx, (index_type, col_types[0]), {}
    )
    fixed_index_typ = fixed_index_sig.return_type

    df_type = DataFrameType(
        col_types,
        fixed_index_typ,
        fake_col_names,
        column_loc=column_loc
    )

    def pd_dataframe_apply_impl(df, udf, args, real_col_names):
        """core dataframe apply implementation logic"""
        row_len = len(df)
        col_len = len(real_col_names)
        lst_col = []
        for _ in range(col_len):
            lst_col.append(np.empty(row_len))
        for row_inx in prange(row_len):
            """return a pd.Series in Mars, in default"""
            udf_arr = udf(df.iloc[row_inx], *args, real_col_names).values
            for col_inx in prange(col_len):
                lst_col[col_inx][row_inx] = udf_arr[col_inx]

        return init_dataframe_internal((lst_col,), df.index, df_type)

    return pd_dataframe_apply_impl

```